bluesky/callbacks/tiled_writer.py

```diff
@@ -0,0 +1,206 @@
+from collections.abc import Iterable
+
+import numpy as np
+import pandas as pd
+from event_model import DocumentRouter, RunRouter
+from pydantic.utils import deep_update
+from tiled.client import from_profile, from_uri
+from tiled.structures.array import ArrayStructure, BuiltinDtype
+from tiled.structures.core import Spec, StructureFamily
+from tiled.structures.data_source import Asset, DataSource, Management
+from tiled.structures.table import TableStructure
+
+MIMETYPE_LOOKUP = {
+    "hdf5": "application/x-hdf5",
+    "ADHDF5_SWMR_STREAM": "application/x-hdf5",
+    "AD_HDF5_SWMR_SLICE": "application/x-hdf5",
+}
+
+
+class TiledWriter:
+    "Write metadata and data from Bluesky documents into Tiled."
+
+    def __init__(self, client):
+        self.client = client
+        self._run_router = RunRouter([self._factory])
+
+    def _factory(self, name, doc):
+        return [_RunWriter(self.client)], []
+
+    @classmethod
+    def from_uri(cls, uri, **kwargs):
+        client = from_uri(uri, **kwargs)
+        return cls(client)
+
+    @classmethod
+    def from_profile(cls, profile, **kwargs):
+        client = from_profile(profile, **kwargs)
+        return cls(client)
+
+    def __call__(self, name, doc):
+        self._run_router(name, doc)
+
+
+class _RunWriter(DocumentRouter):
+    "Write the document from one Bluesky Run into Tiled."
+
+    def __init__(self, client):
+        self.client = client
+        self.start_stop_node = None
```

Comment:
Maybe `run_node` or `root_node` here? Just an idea, not a strong opinion.

Reply:
oh, `root_node` is so much better, thanks!